### PR TITLE
Grpo & verifiable rewards dataset

### PR DESCRIPTION
**Pull request by ianbarber**
#### Context
What is the purpose of this PR? Is it to
- [ x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
* Add GRPO finetuning recipe with LoRA and full model support
* Add verifiable rewards dataset (allows additional columns to pass through, for use in generation)

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

This isn't really mergable as is - I had (as I'm sure others had!) hacked something up over the weekend, and thought it might be worth cleaning up into a more normal format, so sharing to get feedback/see if there is interest in adding a recipe. 

Still running through trying out the recipe, but figured I'd get a steer directionally on if there is interest in adding such a recipe / feedback etc.

### CODE REVIEW

**Comment by ianbarber** [comment]
I looked over #2326 which opened up around the same time, and by my read it's a more complete take and better factored out. Since either will need a bunch of cleanup I"m going to close this one to help focus on landing one version, but happy to follow up and add a LoRA variant etc. or any other bits of cleanup!
